lib/stellar/sep10.rb

```diff
@@ -0,0 +1,353 @@
+        )
+      end
+
+      if !verify_tx_signed_by(transaction_envelope: envelope, keypair: server)
+        raise InvalidSep10ChallengeError.new(
+          "The transaction is not signed by the server"
+        )
+      end
+
+      time_bounds = transaction.time_bounds
+      now = Time.now.to_i
+
+      if time_bounds.nil? || !now.between?(time_bounds.min_time, time_bounds.max_time)
+        raise InvalidSep10ChallengeError.new("The transaction has expired")        
+      end
+
+      # Mirror the return type of the other SDK's and return a string
+      client_kp = Stellar::KeyPair.from_public_key(client_account_id.ed25519!)
+
+      return envelope, client_kp.address
+    end
+
+
+    Contract(C::KeywordArgs[
+      challenge: String,
+      server: Stellar::KeyPair,
+      signers: ArrayOf[Stellar::AccountSigner]
+    ] => C::ArrayOf[Stellar::AccountSigner])
+    # Verifies that for a SEP 10 challenge transaction all signatures on the transaction are accounted for.
+    #
+    # A transaction is verified if it is signed by the server account, and all other signatures match a signer 
+    # that has been provided as an argument. Additional signers can be provided that do not have a signature, 
+    # but all signatures must be matched to a signer for verification to succeed. 
+    #
+    # If verification succeeds a list of signers that were found is returned, excluding the server account ID.
+    #
+    # @param challenge [String] SEP0010 transaction challenge transaction in base64.
+    # @param server [Stellar::Keypair] keypair for server's account.
+    # @param signers [ArrayOf[Stellar::AccountSigner]] The signers of client account.
+    #
+    # @return [ArrayOf[Stellar::AccountSigner]]
+    #
+    # Raises a InvalidSep10ChallengeError if:
+    #     - The transaction is invalid according to Stellar::SEP10.read_challenge_transaction.
+    #     - One or more signatures in the transaction are not identifiable as the server account or one of the 
+    #       signers provided in the arguments.
+    def self.verify_challenge_transaction_signers(
+      challenge:,
+      server:,
+      signers:
+    )
+      if signers.empty?
+        raise InvalidSep10ChallengeError.new("No signers provided.")
+      end
+
+      te, _ = read_challenge_tx(
+        challenge: challenge, server: server
+      )
+
+      # Remove the server signer from the signers list if it is present. It's
+      # important when verifying signers of a challenge transaction that we only
+      # verify and return client signers. If an account has the server as a
+      # signer the server should not play a part in the authentication of the
+      # client.
+      client_signers = Array.new
+      signers.each do |signer|
+        if signer.address != server.address
+          client_signers.push(signer)
+        end
+      end
+
+      signers_found = verify_transaction_signatures(
+        transaction_envelope: te, signers: client_signers
```

Comment:
We need to include the server signer again (I know it's weird we removed it now we're adding it back in, there's a reason for that) so that when we're verifying all the signatures we verify that each signature matches only once to one signer, and that each signer is only consumed once. We can only do that if we include the server signer and match it again along with the client signers.

The reason we remove the server signer earlier is to ensure we have at least one client signer and so that we know what our client signers are separate to our server signer.

See the [Go SDK](https://github.com/stellar/go/blob/c41b3e4309822abc5610230266fc51134ee853fb/txnbuild/transaction.go#L593-L622).